src/basics.ml

```diff
@@ -239,3 +258,120 @@ let term_of_rhs : rule -> term = fun r ->
+  let succ = function
+    | []      -> 0 :: init
+    | x :: xs -> succ x :: xs
+
+  (** [prefix p q] sets position [p] as prefix of position [q], for instance,
+      {i prefix 1 3.4} is {i 1.3.4}; which is represented [prefix [1] [4;3]]
+      is [[4;3;1]]. *)
+  let prefix : t -> t -> t = fun p q -> q @ p
+
+  (** [sub p] returns the position of the first subterm of [p]. *)
+  let sub : t -> t = fun p -> 0 :: p
+
+  (** [tag ?empty s] attaches the positions to a sequence of terms [s] as if
+      they were the subterms of a same term.  If [?empty] is supplied, the
+      first element of the list is at position [empty]. *)
+  let tag : ?empty:t -> term Seq.t -> (term * t) Seq.t = fun ?empty xs ->
+    let start, p = match empty with
+      | None | Some([]) -> 0, init
+      | Some(s :: p)    -> s, p in
+    Seq.mapi (fun i e -> (e, prefix p [i + start])) xs
+end
+
+(** Functional map with [Subterm.t] as keys *)
+module SubtMap = Map.Make(Subterm)
+
+(** {3 Operators on trees} *)
+
+(** [iter l n f t] is a generic iterator on trees; with
+    - function [l] performed on leaves,
+    - function [n] performed on nodes,
+    - [f] returned in case of {!constructor:Fail} on tree [t]. *)
+let tree_iter :
+  do_leaf:((int * int) list -> (term_env, term) Bindlib.mbinder -> 'a) ->
+  do_node:(int -> bool -> 'a TcMap.t -> 'a option -> 'a) ->
+  fail:'a -> tree -> 'a = fun ~do_leaf ~do_node ~fail t ->
+    let rec loop = function
+      | Leaf(pa, a)                                 -> do_leaf pa a
+      | Fail                                        -> fail
+      | Node({ swap ; store ; children ; default }) ->
+          do_node swap store
+            (TcMap.map loop children)
+            (Option.map loop default) in
+    loop t
+
+(** [capacity t] computes the capacity of tree [t].  During evaluation, some
+    terms that are being filtered by the patterns have to be saved in order to
+    be bound in the right hand side of the rule.  The capacity is an upper
+    bound of the number of terms to be saved.  Let [P] be the set of all paths
+    from root to leaves in a tree [t].  Let [s: P → N] the function mapping to
+    any path the number of nodes that have the {!field:store} tag to [true].
+    Then the capacity [c] of [t] is [c = max{s(p) | p ∈ P}]. *)
+let capacity : tree -> int =
+  let do_leaf _ _ = 0 in
+  let fail = 0 in
+  let do_node _ st ch de =
+    let _, chdepths = List.split (TcMap.bindings ch) in
+    let dedepth = Option.get de 0 in
+    List.extremum (>) (dedepth :: chdepths) + (if st then 1 else 0) in
+  tree_iter ~do_leaf:do_leaf ~fail:fail ~do_node:do_node
+
+(** {3 Tree constructor conversion} *)
+
+(** [is_treecons t] returns whether a term [t] is considered as a
+     constructor. *)
+let rec is_treecons : term -> bool = function
+  | Appl(u, _)    -> is_treecons u
+  | Meta(_, _)
+  | Abst(_, _)
+  | Patt(_, _, _) -> false
+  | Vari(_)
+  | Symb(_, _)    -> true
+  | _             -> assert false
+
```

Comment:
You could as well call get_args and check whether the head is Vari or Symb.

Reply:
What happens if it is a Vari? Do you really handle that case too?

Reply:
is_treecons and treecons_of_term do not have the same behavior and the same kind of definition. It should be uniformized: is_treecons should be defined using get_args.

Reply:
Regarding the get_args, it has been modified in future branches.  The Vari(_) case is indeed not truly handled here.